### PR TITLE
Change the inclusivity of exponential histogram bounds

### PR DESCRIPTION
**Pull request by jmacd**
This applies the change specified in https://github.com/open-telemetry/opentelemetry-specification/pull/2633.
- Exactness: implementations SHOULD be correct for exact powers of two
- Exact powers-of-two now fall into the adjacent bucket having one smaller index number.

Part of #3027. 
Follows #2501.

See the surrounding discussion: https://github.com/open-telemetry/opentelemetry-specification/pull/2611.

### CODE REVIEW

**Comment by jmacd** [comment]
Reviewers, there was a [recent change in the specification.](https://github.com/open-telemetry/opentelemetry-specification/pull/2633) This applies the corresponding change to the public exponential histogram mapping functions. PTAL.

File: sdk/metric/aggregator/exponential/mapping/exponent/float64.go [hunk 1]
```diff
@@ -55,15 +55,18 @@ const (
...
 
-// getBase2 extracts the normalized base-2 fractional exponent.  Like
-// math.Frexp(), rounds subnormal values up to the minimum normal
-// value.  Unlike Frexp(), this returns k for the equation f x 2**k
-// where f is in the range [1, 2).
-func getBase2(value float64) int32 {
-	if value <= MinValue {
-		return MinNormalExponent
-	}
+// getBase2 extracts the normalized base-2 fractional exponent.
```
Comment (Aneurysm9) [inline sdk/metric/aggregator/exponential/mapping/exponent/float64.go:58]:
```suggestion
// getNormalBase2 extracts the normalized base-2 fractional exponent.
```

File: sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go [hunk 2]
```diff
@@ -169,3 +188,17 @@ func (l *logarithmMapping) LowerBoundary(index int32) (float64, error) {
 func (l *logarithmMapping) Scale() int32 {
 	return l.scale
 }
+
+// getSignificand returns the 52 bit (unsigned) significand as a
+// signed value.
+func getSignificand(value float64) int64 {
+	return int64(math.Float64bits(value)) & exponent.SignificandMask
```
Comment (Aneurysm9) [inline sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go:195]: This is repeated from the `exponent` package and depends on a value from there.  Should it be exported and reused rather than reimplemented?

Comment (jmacd) [reply sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go:195]: Maybe an internal package?
I'm not sure how breaking it would be, but a bunch of constants in the exponent mapping package do not need to be exported, but they were exported so that ../logarithm could see them. Would you object if I moved everything inessential into an internal package, even if it meant breaking a user who had referred to `exponent.SignificandWidth` for example?

**Review by dashpole — APPROVED** [review]

File: sdk/metric/aggregator/exponential/benchmark_test.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping/exponent"
+	"go.opentelemetry.io/otel/sdk/metric/aggregator/exponential/mapping/logarithm"
+)
+
+func benchmarkMapping(b *testing.B, name string, mapper mapping.Mapping) {
+	b.Run(fmt.Sprintf("mapping_%s", name), func(b *testing.B) {
+		src := rand.New(rand.NewSource(54979))
+
```
Comment (hanyuancheung) [inline sdk/metric/aggregator/exponential/benchmark_test.go:30]: invalid blank line. Remove.

File: sdk/metric/aggregator/exponential/benchmark_test.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+		for i := 0; i < b.N; i++ {
+			_ = mapper.MapToIndex(1 + src.Float64())
+		}
+	})
+}
+
+func benchmarkBoundary(b *testing.B, name string, mapper mapping.Mapping) {
+	b.Run(fmt.Sprintf("boundary_%s", name), func(b *testing.B) {
+		src := rand.New(rand.NewSource(54979))
+
```
Comment (hanyuancheung) [inline sdk/metric/aggregator/exponential/benchmark_test.go:40]: Ditto #L30

**Review by hanyuancheung — APPROVED** [review]

File: sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go (hunk 2 above)
Comment (Aneurysm9) [reply sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go:195]: These aren't stable packages, are they?  I'm fine with making breaking changes if they improve the interface.  I guess the question I'd have would be why an external user would be referring to that constant and whether that need would continue and force them to assume/duplicate it or is it eliminated by changes made here or that could be made.

File: sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go [hunk 5]
```diff
@@ -212,11 +236,14 @@ func TestExponentMappingNegFour(t *testing.T) {
 		{0x1p-1023, -64},
```
Comment (jack-berg) [inline sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go:236]: Some of these min / subnormal buckets appear to be incorrect. For example, I believe `0x1p-1074,` should be in the -68 bucket instead of -64.

Am I missing something or is the implementation ignoring perfect precision around subnormal values intentionally?

Comment (jmacd) [reply sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go:236]: I believe it is a mistake to treat subnormal values as real values.

One of the properties that I wanted to enforce in the data structure is as follows:
- When a bucket B1 is calculated at scale S1
- And a bucket B2 calculated for the same value at scale S2,
- Where S2 < S1
- We expect (B1 >> (S1-S2)) == B2, which follows from the "perfect subsetting"

The simple logarithm mapping functions are not accurate in the subnormal range. Rounding values up to the smallest normal value ensures the property above. Imagine a histogram where the first entered value is 0x1p-1050 for example. With one value, we can use the maximum scale, which is 20 in this case. However the logarithm implementation does not function for subnnormals, thus to ensure that the bucket computed at scale 20 matches the one for negative scales, we round up subnormals.

I would like to reserve the right to improve both mapping functions -- as long as the last property holds the data structure will not be impacted by the decision to support subnormals.

Comment (jack-berg) [reply sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go:236]: That makes sense. But the [implementation from the spec](https://github.com/open-telemetry/opentelemetry-specification/pull/2633/files#diff-6c97e08ad12fe1538de64c06f0cab436e9115287ea35183b88dbd1df77ce0bc5R745-R759) has special logic accommodating subnormal values which applies to scales <= 0 since negative scales are just a [bitshift of the zero scale algorithm](https://github.com/open-telemetry/opentelemetry-specification/pull/2633/files#diff-6c97e08ad12fe1538de64c06f0cab436e9115287ea35183b88dbd1df77ce0bc5R803-R805).

File: sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go (hunk 2 above)
Comment (jmacd) [reply sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go:195]: I agree. I'll update this PR to make things internal.

File: sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go (hunk 5 above)
Comment (jmacd) [reply sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go:236]: Yes. Although I wrote that (with @reyang's help 😂), I followed it with the [another example allowing implementations to round subnormal values](https://github.com/open-telemetry/opentelemetry-specification/pull/2633/files#diff-6c97e08ad12fe1538de64c06f0cab436e9115287ea35183b88dbd1df77ce0bc5R772-R790). The requirement being imposed by me here is that the two mapping functions handle subnormals the same way, which does not feel like a spec-level requirement, #3022 requires the two mapping functions to be consistent.

TL;DR Another reason to avoid treating subnormals, IMO, is that it changes the total maximum range of the data structure in addition to making it less symmetric.  Normal exponents range from [-1022, 1023], which fits nicely into 11 bits of information. If you allow subnormal values, the range is [-1074, 1023] which requires 12 bits of information (with most of 1 bit unused). Since we are limited to 32 bits of index (in OTLP), taking 12 bits of information for the index would decrease the maximum and minimum scale parameters used in #3022. MaxScale=20 implies 20 bits of index per exponent, so with only normal values MaxScale=20 plus 11 bits of exponent fits a 32 bit integer. Allowing subnormals would limit MaxScale to 19 and require MinScale=-11.

**Comment by jmacd** [comment]
(Will be a few days before this is updated.)

File: sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go (hunk 5 above)
Comment (reyang) [reply sdk/metric/aggregator/exponential/mapping/exponent/exponent_test.go:236]: Here are some test cases I wrote for OpenTelemetry.NET just FYI https://github.com/open-telemetry/opentelemetry-dotnet/blob/245c2b3e29a350e3cfef2168dcb29be1fcda38f8/test/OpenTelemetry.Tests/Metrics/ExponentialBucketHistogramTest.cs#L146.

File: sdk/metric/aggregator/exponential/benchmark_test.go (hunk 4 above)
Comment (jmacd) [reply sdk/metric/aggregator/exponential/benchmark_test.go]: Fixed.

File: sdk/metric/aggregator/exponential/benchmark_test.go (hunk 3 above)
Comment (jmacd) [reply sdk/metric/aggregator/exponential/benchmark_test.go]: Fixed.

File: sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go (hunk 2 above)
Comment (jmacd) [reply sdk/metric/aggregator/exponential/mapping/logarithm/logarithm.go]: Fixed. The two helper functions and all the internal constants are in an `./internal` package now.

**Comment by jmacd** [comment]
LowerBoundary() needs to be exact to match the the mapping function, will resolve.
